python/lsst/ts/dsm/dsm_csc.py

```diff
@@ -1,34 +1,351 @@
         self.telemetry_directory = None
         self.telemetry_loop_running = False
         self.telemetry_loop_task = None
-        super().__init__("DSM", index, initial_state=initial_state,
+        self.telemetry_watcher = aionotify.Watcher()
+        self.simulated_telemetry_ui_config_written = False
+        self.simulated_telemetry_loop_running = False
+        self.simulated_telemetry_task = None
+        self.simulation_loop_time = None
+        self.config = None
+
+        self.loop_die_timeout = 5  # how long to wait for the loops to die?
+
+        super().__init__("DSM", index, schema_path=schema_path, config_dir=config_dir,
+                         initial_state=initial_state,
                          initial_simulation_mode=initial_simulation_mode)
 
         ch = logging.StreamHandler()
         self.log.addHandler(ch)
 
-    async def implement_simulation_mode(self, simulation_mode):
-        if simulation_mode:
+    async def begin_exitControl(self, id_data):
+        """Begin do_exitControl; called after state changes but before command
+           acknowledged.
+
+        This method will remove the telemetry directory if in simulation mode.
+
+        Parameters
+        ----------
+        id_data : `CommandIdData`
+            Command ID and data
+        """
+        if self.simulation_mode:
+            if os.path.exists(self.telemetry_directory):
+                os.removedirs(self.telemetry_directory)
+
+    async def begin_standby(self, id_data):
+        """Begin do_standby; called after state changes but before command
+           acknowledged.
+
+        This method will stop the telemetry and if necessary the simulation
+        loops.
+
+        Parameters
+        ----------
+        id_data : `CommandIdData`
+            Command ID and data
+        """
+        self.simulated_telemetry_loop_running = False
+        self.telemetry_loop_running = False
+
+    async def begin_start(self, id_data):
+        """Begin do_start; called after state changes but before command
+           acknowledged.
+
+        This method will create the telemetry directory if in simulation mode
+        and the directory has not been specified previously.
+
+        Parameters
+        ----------
+        id_data : `CommandIdData`
+            Command ID and data
+        """
+        if self.simulation_mode and self.telemetry_directory is None:
             self.telemetry_directory = tempfile.mkdtemp()
-            self.log.info(f"Creating temporary directory: {self.telemetry_directory}")
+            self.log.debug(f"Creating temporary directory: {self.telemetry_directory}")
+
+        await super().begin_start(id_data)
+
+    def cleanup_simulation(self):
+        """Remove all generated files and directory from simulation
+        """
+        if os.path.exists(self.telemetry_directory):
+            for tfile in os.listdir(self.telemetry_directory):
+                os.remove(os.path.join(self.telemetry_directory, tfile))
```

Comment:
This has turned out to be a bigger problem than expected if I put lots of code in `report_summary_state`. I'd like some time to discuss this with you.

Reply:
I looked at your new code in `report_summary_state` and it looks fine to me. You might consider logging at debug level instead of info and once you trust the code you can consider dropping the log messages entirely, at which point you can replace this:
```
if self.simulation_mode and not self.simulated_telemetry_loop_task.done():
    self.log.info("Shutting down simulated telemetry loop.")
    self.simulated_telemetry_loop_task.cancel()
    self.simulated_telemetry_ui_config_written = False

    self.log.info("Cleanup simulation files.")
    self.cleanup_simulation()

if not self.telemetry_loop_task.done():
    self.log.info("Shutting down telemetry loop.")
    self.telemetry_loop_task.cancel()
```
with this simpler block, assuming cleanup_simulation is always safe to call:
```
if self.simulation_mode:
    self.simulated_telemetry_loop_task.cancel()
    self.simulated_telemetry_ui_config_written = False
    self.cleanup_simulation()

self.telemetry_loop_task.cancel()
```
because a task can safely be canceled in any state (it is a no-op if the task is done).

I suggest moving the code from `begin_exitControl` to method `async def close_tasks(self)` which is the standard way to cleanup at shutdown. That method can also cancel all background tasks if you want to be paranoid, though the code in report_summary_state probably suffices for that.